### PR TITLE
Feature: suport public ip auto detected

### PR DESCRIPTION
**Pull request by njucjc**

### CODE REVIEW

File: pkg/k8s/engine_controller.go [hunk 1]
```diff
@@ -149,6 +155,32 @@ func (c *EngineController) shouldHandleGateway(gateway *v1alpha1.Gateway) bool {
...
+		return nil
+	}
+	publicIP, err := utils.GetPublicIP()
+	if err != nil {
+		return err
+	}
+	// retry to update ip
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var i int
+		for i = 0; i < len(gateway.Spec.Endpoints); i++ {
```
Comment (DrmagicE) [inline pkg/k8s/engine_controller.go:169]: How about using `for range` here and putting the update logic inside the iteration. Like this:
```
        // retry to update ip
	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
		for k, v := range gateway.Spec.Endpoints {
			if v.NodeName == c.nodeName {
				gateway.Spec.Endpoints[k].PublicIP = publicIP
				_, err = c.ravenClient.RavenV1alpha1().Gateways().Update(context.Background(), gateway, v1.UpdateOptions{})
				return err
			}
		}
		return nil
	})
```

File: pkg/k8s/engine_controller.go [hunk 2]
```diff
@@ -149,6 +155,32 @@ func (c *EngineController) shouldHandleGateway(gateway *v1alpha1.Gateway) bool {
...
+func (c *EngineController) completeGateway(gateway *v1alpha1.Gateway) error {
+	if gateway.Status.ActiveEndpoint.NodeName != c.nodeName {
+		return nil
+	}
+	publicIP, err := utils.GetPublicIP()
+	if err != nil {
+		return err
+	}
+	// retry to update ip
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
Comment (DrmagicE) [inline pkg/k8s/engine_controller.go:167]: I think we should get the gateway Object from the APIServer on each retry before attempting to update the gateway.

Comment (njucjc) [reply pkg/k8s/engine_controller.go:167]: fixed

File: pkg/k8s/engine_controller.go (hunk 1 above)
Comment (njucjc) [reply pkg/k8s/engine_controller.go]: done

File: pkg/k8s/engine_controller.go [hunk 3]
```diff
@@ -149,6 +155,33 @@ func (c *EngineController) shouldHandleGateway(gateway *v1alpha1.Gateway) bool {
...
+	if gateway.Status.ActiveEndpoint.NodeName != c.nodeName {
+		return nil
+	}
+	publicIP, err := utils.GetPublicIP()
+	if err != nil {
+		return err
+	}
+	// retry to update public ip of gateway
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		// get gateway for api server
```
Comment (DrmagicE) [inline pkg/k8s/engine_controller.go:168]: for => from?

Comment (njucjc) [reply pkg/k8s/engine_controller.go]: fixed

**Comment by DrmagicE** [comment]
lgtm. @BSWANG  Do you have any comments?

**Comment by BSWANG** [comment]
> lgtm. @BSWANG Do you have any comments?

lgtm

**Pull request merged**